### ISSUE TITLE
Website search usability issues

### ISSUE THREAD

**Issue by mattlord** [vitessio/website#977]
The current website search (Algolia) is of limited practical use today for the following reasons:
1. The results include no information to identify which specific page it would take you to (e.g. reference or user guide for vtgate)
2. We now have versioned docs so the same page is repeated N times in the results (and because of issue 1, you cannot see which version a result is for). When combined with the fixed number of max results shown, you often cannot get to what you want

We should see if we can add any sort of preview or metadata to the search results so that you can see where it will take you w/o having  to click on it. 

We should see if we can default to ONLY searching the [`$current` version](https://github.com/vitessio/website/blob/098e06a996c4359bd79b321f07f2966a498e1acf/config.toml#L16) of the docs.

**Comment by nate-double-u** [comment]
I opened two PRs on etcd.io's website repo a while ago relating to search:
* https://github.com/etcd-io/website/pull/403
* https://github.com/etcd-io/website/pull/392

Their site is set up a little differently to this one, but Lunr or Google may be a good solution to your search issues. You can see how the search works on the https://etcd.io site. Since vitess.io isn't using docsy it may be different to set up, but this may be a good starting point.

**Comment by chalin** [comment]
@mattlord, would a switch to Lunr be an acceptable solution?

**Comment by mattlord** [comment]
Hi @chalin! 

Yeah, I don't think we have any real preferences on the implementation used for vitess.io. Any pointers on how we'd do that? 

I guess we'd glean that from looking at the two noted examples where etcd did this?
* https://github.com/etcd-io/website/pull/403
* https://github.com/etcd-io/website/pull/392

Are there any other suggestions, requirements, ideas on the CNCF side that we should know about?

Thanks again!

**Comment by chalin** [comment]
Hi @mattlord - no, nothing else comes to mind. If the Vitess website was using the [Docsy](https://github.com/google/docsy) theme, then the switch to Lunr would be simple and quick to do. While that isn't the case, you can always look at the Docsy implementation for inspiration. (I'd be glad to lend a hand, but it looks like it wouldn't be before the start of 22Q3.)

**Comment by derekperkins** [comment]
Who controls our Algolia account? Is it @lucperkins or the CNCF? It shouldn't be too hard to configure Algolia to do that correctly.

**Comment by deepthi** [comment]
@chalin @nate-double-u do you know the answer to the algolia question?

**Comment by nate-double-u** [comment]
Hi @deepthi, I'm not sure who controls the account -- I've looked around our docs to see if there are any records but I haven't found any.

If you're on the free tier tho, I think the Algolia `docsearch` widget is controlled by this file: https://github.com/algolia/docsearch-configs/blob/master/configs/vitess.json, which I think you can open a PR against.

The only site I'm aware of that has done versioning in their search correcly is [velero.io](https://velero.io/docs/v1.9/), and for compairison their `docsearch` file looks like this: https://github.com/algolia/docsearch-configs/blob/master/configs/velero.json

**Comment by mattlord** [comment]
Thanks, @nate-double-u ! I confirmed that the version based search works intuitively and well on https://velero.io/docs/

So it seems to be handled this way:
1. Sending algolia the version in each page:
```
...
            <button class="btn btn-primary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown"
                    aria-haspopup="true" aria-expanded="false">
                v1.6
            </button>
...
            
<script type="text/javascript"> docsearch({
    apiKey: '****',
    indexName: 'velero',
    inputSelector: '.docsearch-input',
    algoliaOptions: {'facetFilters': ["version:v1.6"]},
    debug: false 
  });
</script>
</body>

</html>
```
2. Telling algolia to use that version facet:
```
  "selectors": {
...
    "version": {
      "selector": "#dropdownMenuButton",
      "global": true,
      "default_value": "none"
    }
  },
  "strip_chars": " .,;:#",
  "custom_settings": {
    "attributesForFaceting": [
      "language",
      "version"
    ]
  },
```

This is really helpful! :heart: I'm not sure that there's a good way to test it w/o modying the prod site but we can give it a shot. Unfortunately the search is of limited use today given the many versions we have now so that gives us some leeway to play with it. 🙂

**Comment by nate-double-u** [comment]
You can see some more of how they managed it here:
* https://github.com/vmware-tanzu/velero/blob/e7a1c3d4dc38b2761f9c78913df0f2e2ed6e866f/site/layouts/docs/docs.html
* https://github.com/vmware-tanzu/velero/blob/e9c997839ebbb04688a88ffb981eeb74de51b380/site/assets/_scss/site/layouts/_docsearch.scss

Vitess' equivalent is here:
* https://github.com/vitessio/website/blob/bc636cb69e3b414340deb62b556ec2b2c3b68735/layouts/partials/javascript.html (and you have your apiKey here as well -- i'm not sure if that's helpful or not)

**Comment by deepthi** [comment]
I thought we could just change the config in the `docsearch-configs` repo by sending a PR but it is archived!!
We have to do a migration as documented here https://docsearch.algolia.com/docs/migrating-from-legacy/

**Comment by deepthi** [comment]
@nate-double-u @derekperkins do you know whether doing the migration to the "new" way of configuring Algolia requires an account? AFAIK we have neither a vitess account nor a CNCF one.

**Comment by nate-double-u** [comment]
I don't know (I don't think there is a CNCF controlled one).

Depending on how much work the migration is, it may be worth looking into other alternatives to see if they are as much/less work to implement.

**Comment by llibdude** [comment]
From https://github.com/algolia/docsearch-configs
`This repository has been archived by the owner. It is now read-only.`

![image](https://user-images.githubusercontent.com/5255924/174880111-8820ece5-d71a-4377-93c9-e607fb212e8a.png)

I'm not sure if I'm interpretting this screenshot correctly but the last line `If you have not joined your new application yet, please check your emails!` might suggest an email was sent out when this repo was archived perhaps with our account details.

I don't know what email might have been used when we originally signed up but if anyone can search any emails they have access to perhaps they can find something?


I don't think using algoria would be any harder/easier than any other options and we can stick with it assuming we can either recover or create our account.  Maybe someone can also just reach out directly to algoria for support on this.

**Comment by deepthi** [comment]
I'm going to request an algolia account for Vitess. It seems (from talking to some others) that it should be easy to setup if we have an account

**Comment by deepthi** [comment]
Created a CNCF service desk issue https://cncfservicedesk.atlassian.net/servicedesk/customer/portal/1/CNCFSD-1314

**Comment by deepthi** [comment]
I went ahead and created an algolia account for domain `vitess.io`.
Invited @llibdude @derekperkins and @mattlord to it. Check your work emails 😄 
@nate-double-u should I invite you as well?

**Comment by llibdude** [comment]
@deepthi sorry I think I may have misunderstood Algolia's products.  We don't want a normal Algolia account after all.  We don't want to care about managing our own index, etc and instead want to leverage the auto scraping doc gen.

As was discussed previously we want to upgrade our DocSearch to v3.  To do that I believe we would need to reapply for DocSearch https://docsearch.algolia.com/apply/ instead of a normal Algolia account.

I have reached out to support to clarify.  Hopefully they respond today but if not I think we should reapply to docsearch and see what happens.

**Comment by llibdude** [comment]
A colleague helped me understand more about how to set up the docsearch scraper with our own Algolia account. I'm still curious as to the response from support since having search managed for us in this project may be advantageous to needing to manage yet another account.

**Comment by nate-double-u** [comment]
> I went ahead and created an algolia account for domain `vitess.io`.
> 
> Invited @llibdude @derekperkins and @mattlord to it. Check your work emails 😄 
> 
> @nate-double-u should I invite you as well?

Yes please--sorry for the delay in responding

**Comment by llibdude** [comment]
Hey all, check out this PR I crafted https://github.com/vitessio/website/pull/1070. I wrote up a description in the PR along with a video to prove it's working.  There are a few minor TODOs (set up a paid account, set up github secrets) but otherwise will work.

We have to decide if we just want to go this route or see about working with Algolia to get a potentially completely free account set up on their v3.  I will reach out to them again next week but figured I'd have the PR ready in case we need it.
